File: Baekjoon/1967_트리의지름.py
import sys
from collections import deque


def input():
    return sys.stdin.readline().rstrip()


N = int(input())
tree = [[] for _ in range(N + 1)]


def dfs(node, total):
    q = deque([[node, total]])
    while q:
        node, total = q.popleft()
        for i in tree[node]:
            child, cost = i
            if visited[child] == -1:
                visited[child] = cost + total
                q.append([child, cost + total])


def reset_visited(idx):
    visited = [-1] * (N + 1)
    visited[idx] = 0

    return visited


tree = [[] for _ in range(N + 1)]

for _ in range(N - 1):
    a, b, c = map(int, input().split())
    tree[a].append([b, c])
    tree[b].append([a, c])

visited = reset_visited(1)
visited[1] = 0
dfs(1, 0)

node1 = visited.index(max(visited))
visited = reset_visited(node1)
dfs(node1, 0)

print(max(visited))

File: Baekjoon/1967_트리의지름.js
const fs = require('fs');
const input = fs.readFileSync('./dev/stdin').toString().trim();

function main(input) {
  if (typeof input !== 'string' || !input) return;

  const [stringN, ...inputArr] = input.split('\n');

  const N = Number(stringN);

  // 부모노드 자식노드 가중치
  const tree = inputArr.reduce(
    (tree, line) => {
      const [parent, child, weight] = line.split(' ').map(Number);

      tree[parent].push([child, weight]);
      tree[child].push([parent, weight]);

      return tree;
    },
    Array.from({ length: N + 1 }, () => [])
  );

  const firstBFSNode = bfs(tree, 1, N + 1);

  return bfs(tree, firstBFSNode.node, N + 1).weight;
}

function bfs(arr, s, N) {
  const q = [[s, 0]];
  const visited = Array(N).fill(false);
  const maxLengthNodeObj = {
    node: 0,
    weight: 0,
    update(node, weight) {
      if (this.weight >= weight) return;

      this.weight = weight;
      this.node = node;
    },
  };

  while (q.length > 0) {
    const [node, weight] = q.shift();

    if (visited[node]) continue;

    visited[node] = true;
    maxLengthNodeObj.update(node, weight);

    arr[node]?.forEach(([childNode, childWeight]) => {
      q.push([childNode, childWeight + weight]);
    });
  }

  return maxLengthNodeObj;
}

console.log(main(input));

module.exports = main;
